Improve coverage for fs operations

mem opened the pull request:
* TestCopyDir: add subdirectory to the test case; extend test to handle
  multiple files in different directories.

* Add tests for error conditions in CopyDir: source directory cannot be
  read; destination directory cannot be read; files cannot be read.

* Add tests for error conditions in IsRegular.

* Add tests for error conditions in IsDir.

This addresses issue #41.

Signed-off-by: Marcelo E. Magallon <marcelo.magallon@gmail.com>

mem commented:
Interesting, Windows behaves differently wrt to the errors that are produced. I guess setting the file/directory permissions doesn't work / doesn't work as expected on that platform.

sdboyer commented:
This would now need to be refactored to incorporate the new test harness - https://github.com/golang/dep/blob/master/cmd/dep/testdata/harness_tests/README.md. As this PR has been languishing for a while now, I'm going to close it - please feel free to reopen a new PR with the same basic goal!

sdboyer commented:
bump

sdboyer commented:
Maybe @mattn or @tro3 could lend some thoughts on how to make this work for windows? I know both of you fine folks dev on that OS 😄

sdboyer commented on fs_test.go:
```diff
@@ -25,52 +26,164 @@ func TestCopyDir(t *testing.T) {
+}
+
+func TestCopyDirFailSrc(t *testing.T) {
+	var srcdir, dstdir string
```
@mattn Sure, that's one approach. I wash hoping you might have some thoughts about how to set up a different test on Windows that checks for something analogous. So, instead of `os.Chmod()`...idk, because I don't understand Windows' file permission model.

mem commented:
Sorry about the silence for the last few weeks (new kid :-)

I've updated the PR with the suggestions and I added the skip to the failing test. Hopefully someone with a better grasp of Windows can provide a better implementation in the future.

sdboyer commented:
> Sorry about the silence for the last few weeks (new kid :-)

!👶 👶 !! 🎉!! 🎉🎉! 🎉❤️🎉 🎉🎉❤️ 🎉 🎊 🎊!!!!!!

I'm out of time for dep for tonight, but hopefully will have time to get to reviewing these tomorrow. 😄

davecheney commented:
ping, this potentially blocks #539

sdboyer commented:
Ach yeah, I totally dropped this one, thanks for pinging on it. Even had a partially-written review on it. It was really just about the windows bits, though - the rest of this looked fine.

I've opened a #548 as a follow-up for the windows bits.
